File: my-solutions/0133-clone-graph/solution.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        node_map = {}

        def clone(node):
            if node in node_map:
                return node_map[node]
            else:
                # clone the node
                curr = Node(node.val)

                # add curr to the map
                node_map[node] = curr

                for i in range(len(node.neighbors)):
                    curr.neighbors.append(clone(node.neighbors[i]))
                
                return curr

        return clone(node) if node else None

File: my-solutions/0133-clone-graph/solution.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    Node* cloneGraph(Node* node) {
        if(!node) return nullptr;
        unordered_map<Node*, Node*> nodemap{{node, new Node(node->val)}};
        dfs(node, nodemap);
        return nodemap[node];
    }

    void dfs(Node* node, unordered_map<Node*, Node*>& nodemap) {
        for(auto& n: node->neighbors) {
            if(nodemap.find(n) == nodemap.end()) {
                nodemap[n] = new Node(n->val);
                dfs(n, nodemap);
            }
            nodemap[node]->neighbors.push_back(nodemap[n]);
        }
    }
};
